### DISCUSSION TITLE
Why did brew install ffmpeg update everything else I've ever installed?

### DISCUSSION THREAD

**Discussion by rick-rheo** [Homebrew/discussions#1235] [Everyday usage]
I just needed ffmpeg.  That's all.  Well, brew decided I needed to upgrade mysql, redis, php, haproxy, memcache and a bunch of other stuff before it installed ffmpeg.  ...and it completely destroyed my PHP install and did the same for MySQL.  Both are burnt toast at the moment.  Redis got a major upgrade which I most definitely didn't want because I need to make sure code I build/test will run on the redis 2.8 install in production.  I'm so pissed off.  I just wanted ffmpeg.  ffmpeg does not need redis, mysql, etc. so why in the hell was that stuff touched.    I also needed a version of PHP that's no longer supported anywhere and trying to re-link that version doesn't work. I essentially have a brick for a dev machine right now.

**Comment by SMillerDev** [comment]
https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff

TLDR: if you need specific versions docker is a better fit than an evergreen package manager like brew.

**Reply by rick-rheo** [reply]
i used brew install   not brew upgrade

**Reply by rick-rheo** [reply]
and docker is a battery and memory suck and I often need to be in places where I can't plug in.

**Reply by carlocab** [reply]
The linked FAQ also applies to `brew install`. If you need specific versions and don't want to use Docker, then you need to maintain these versioned formulae in your own tap.

Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

**Reply by rick-rheo** [reply]
BUT why does brew decide to "help" me by upgrading everything else ESPECIALLY stuff that has absolutely nothing to do with what I'm installing?  I just wanted ffmpeg. Why did I even need to consider what this would do to mysql?   apt-get install doesn't do that.   I don't get an installer for some piece of software the goes ahead and upgrades all my other applications.   This is a very very very dumb thing to do.

**Reply by rick-rheo** [reply]
...at the very least WARN that installing one package will upgrade a bunch of others and let me cancel rather than plowing ahead utterly destroying my dev machine

**Reply by carlocab** [reply]
It doesn't upgrade everything. It upgrades everything connected to `ffmpeg` via its dependency graph. It just so happens that `ffmpeg` has an extremely large dependency graph.

If you installed something with no dependencies (try `brew install hello` when you have some outdated formulae), then the install won't upgrade anything.

**Reply by tlk** [reply]
> apt-get install doesn't do that.

Homebrew was changed some years ago to silently auto-update on every(?) command.

When you ran that `brew install ffmpeg` command it silently carried out a `brew update` operation similar to running `apt-get update`. After this it moved on to installing ffmpeg for you, and as carlocab described it found that it was necessary to upgrade a bunch of already installed stuff.

What is particularly confusing is why it decided to upgrade the existing software when all you asked was for it to install an unrelated piece of software. I believe all the unwanted upgrades happened because of the silent `brew update` operation. Because the ffmpeg formula was brand new it triggered an upgrade of some libraries that were already installed, and this triggered a snowball-effect that ended up upgrading the world and its dog.

I personally prefer another default setting but I am not a maintainer.

You may be interested in https://github.com/Homebrew/discussions/discussions/1107#discussioncomment-517248 and https://github.com/Homebrew/discussions/discussions/578#discussioncomment-295595

The keyword you want to look for is HOMEBREW_NO_AUTO_UPDATE.

**Reply by carlocab** [reply]
> The keyword you want to look for is HOMEBREW_NO_AUTO_UPDATE.

I don't recommend doing this. However, if you must, please don't forget to `brew update` at the end of the month. You don't want to get stuck trying to download bottles from a non-existent Bintray.
